__tests__/distributors/base.test.ts

```diff
@@ -0,0 +1,180 @@
+    mockJavaBase = new EmptyJavaBase(input);
+    const mockJavaBasePrototype = Object.getPrototypeOf(mockJavaBase);
+    expect(mockJavaBasePrototype.findInToolcache.call(mockJavaBase)).toEqual(expected);
+  });
+});
+
+describe('Test setupJava', () => {
+  const actualJavaVersion = '11.1.10';
+  const javaPath = path.join('Java_Empty_jdk', actualJavaVersion, 'x86');
+
+  let mockJavaBase: EmptyJavaBase;
+
+  let tcFind: jest.SpyInstance;
+  let coreDebug: jest.SpyInstance;
+  let coreInfo: jest.SpyInstance;
+  let coreExportVariable: jest.SpyInstance;
+  let coreAddPath: jest.SpyInstance;
+  let coreSetOutput: jest.SpyInstance;
+
+  beforeEach(() => {
+    tcFind = jest.spyOn(tc, 'find');
+    tcFind.mockImplementation((toolname: string, javaVersion: string, architecture: string) => {
+      const semverVersion = new semver.Range(javaVersion);
+      const possiblePath = path.join(toolname, javaVersion, architecture);
+
+      if (path.basename(javaPath) !== architecture || !javaPath.includes(toolname)) {
+        return '';
+      }
+
+      return semver.satisfies(actualJavaVersion, semverVersion) ? javaPath : '';
+    });
+
+    // Spy on core methods
+
+    coreDebug = jest.spyOn(core, 'debug');
+    coreDebug.mockImplementation(() => undefined);
+
+    coreInfo = jest.spyOn(core, 'info');
+    coreInfo.mockImplementation(() => undefined);
+
+    coreAddPath = jest.spyOn(core, 'addPath');
+    coreAddPath.mockImplementation(() => undefined);
+
+    coreExportVariable = jest.spyOn(core, 'exportVariable');
+    coreExportVariable.mockImplementation(() => undefined);
+
+    coreSetOutput = jest.spyOn(core, 'setOutput');
+    coreSetOutput.mockImplementation(() => undefined);
+  });
+
+  afterEach(() => {
+    jest.resetAllMocks();
+    jest.clearAllMocks();
+    jest.restoreAllMocks();
+  });
+
+  it.each([
+    [
+      { version: '11', arch: 'x86', packageType: 'jdk' },
+      { javaVersion: actualJavaVersion, javaPath }
+    ],
+    [
+      { version: '11.1', arch: 'x86', packageType: 'jdk' },
+      { javaVersion: actualJavaVersion, javaPath }
+    ],
+    [
+      { version: '11.1.10', arch: 'x86', packageType: 'jdk' },
+      { javaVersion: actualJavaVersion, javaPath }
+    ]
+  ])(
+    'should find java for path %o -> %s, throw an error for non implemented method',
+    (input, expected) => {
+      mockJavaBase = new EmptyJavaBase(input);
+      expect(mockJavaBase.setupJava()).resolves.toEqual(expected);
+      expect(tcFind).toHaveBeenCalled();
+    }
+  );
+
+  it.each([
+    [{ version: '11', arch: 'x86', packageType: 'jre' }],
+    [{ version: '8', arch: 'x86', packageType: 'jdk' }],
+    [{ version: '11', arch: 'x64', packageType: 'jdk' }],
+    [{ version: '11', arch: 'x64', packageType: 'jre' }]
+  ])('should throw an error for non implemented method', async input => {
+    mockJavaBase = new EmptyJavaBase(input);
+    await expect(mockJavaBase.setupJava()).rejects.toThrowError('Method not implemented.');
```

Comment:
I am not sure that it makes much sense to test `setupJava` method since it is pretty straightforward.
I am a bit concerned that actually we use stub class where methods not implemented and this method always throws exception. As a result, we test only begin of this method (before first call of non-implemented method).
Does it really make sense?

Reply:
Done. Thank you.